### DISCUSSION TITLE
Problem with prisma relations after upgrading to 4.11 and referentialIntegrity

### DISCUSSION THREAD

**Discussion by lucasnachtigall** [prisma/prisma#18491] [Q&A]
So I had the following models(changed names to make it more simple to understand I guess)

```
model User {
  id    String @id @default(cuid())
  name  String @db.VarChar(256)
  jobs  Job[]
  tasks Task[]

  @@unique([userId, name])
  @@index([userId])
}

model Job {
  id            String  @id @default(cuid())
  userId        String
  user          User    @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: NoAction)
  currentTaskId String?
  currentTask   Task?   @relation(fields: [currentTaskId], references: [id], onDelete: NoAction, onUpdate: NoAction, name: "currentTask")
  tasks         Task[]  @relation(name: "tasks")

  @@index([userId])
}

model Task {
  id          String @id @default(cuid())
  userId      String
  user        User   @relation(fields: [siteId], references: [id], onDelete: Cascade, onUpdate: NoAction)
  jobId       String
  job         Job    @relation(fields: [jobId], references: [id], onDelete: NoAction, onUpdate: NoAction, name: "tasks")
  currentJobs Job[]  @relation(name: "currentTask")

  @@index([userId])
  @@index([jobId])
}
```

Where: 1 user could have many jobs 1 Job could have many tasks, but each job only has one current Task.

This used to work fine with referentialIntegrity, but now when I try to delete an user I get the following error:

  ```
The change you are trying to make would violate the required relation 'tasks' between the `Task ` and `Job ` models.
  meta: {
    relation_name: 'tasks',
    model_a_name: 'Task',
    model_b_name: 'Job '
  }
  ```

I think this is because some cyclic problem when using relationMode = "prisma", where it tries to delete the Jobs, but those have tasks, and those tasks have jobs.

I tried multiple ways (even asked chatGPT for help, it was clueless too.) but always ended up getting the same error or "Reference causes a cycle" when trying to set onDelete to cascade on both Job and Tasks models...

Hope someone have an idea whats wrong. Thanks in advance!

**Comment by ludralph** [comment]
Hi @lucasnachtigall 👋 

Can you please set the `onDelete` cascade to `Job` and `Task` relations to `SetNull` instead of `Cascade`. This way, when you delete an User record, all related Job records will have their userId field set to null, and all related Task records will have their jobId field set to null as well? Let's see if it behaves correctly with the same version.

**Reply by lucasnachtigall** [reply]
Hello, thanks for the answer

I tried setting chaging those values to SetNull as following:
`model Job {
  id            String  @id @default(cuid())
  userId        String
  user          User    @relation(fields: [userId], references: [id], onDelete: SetNull, onUpdate: NoAction)
  currentTaskId String?
  currentTask   Task?   @relation(fields: [currentTaskId], references: [id], onDelete: NoAction, onUpdate: NoAction, name: "currentTask")
  tasks         Task[]  @relation(name: "tasks")

  @@index([userId])
}

model Task {
  id          String @id @default(cuid())
  userId      String
  user        User   @relation(fields: [siteId], references: [id], onDelete: SetNull, onUpdate: NoAction)
  jobId       String
  job         Job    @relation(fields: [jobId], references: [id], onDelete: NoAction, onUpdate: NoAction, name: "tasks")
  currentJobs Job[]  @relation(name: "currentTask")

  @@index([userId])
  @@index([jobId])
}`

When trying to delete I got the following error: "Null constraint violation on the (not available)"

But I think the main problem is between the "Job" and"Task" since they kinda have a cyclic relation...

**Reply by lucasnachtigall** [reply]
@ludralph , doing some extra research I believe my problem is because of this update on prisma : https://github.com/prisma/prisma/issues/15655, where NoAction = Restrict now. 

Is there any way to keep the Task and Job records even if the User record is deleted?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
